src/NHibernate/Loader/Criteria/CriteriaQueryTranslator.cs

```diff
@@ -408,15 +439,26 @@ private void CreateCriteriaCollectionPersisters()
 
 		private Persister.Entity.IJoinable GetPathJoinable(string path)
 		{
-			NHibernate_Persister_Entity.IJoinable last = (NHibernate_Persister_Entity.IJoinable)Factory.GetEntityPersister(rootEntityName);
-			NHibernate_Persister_Entity.IPropertyMapping lastEntity = (NHibernate_Persister_Entity.IPropertyMapping)last;
-
-			string componentPath = "";
+			// start with the root
+			IJoinable last = rootPersister;
+			
+			var tokens = path.Split(new[] {'.'}, StringSplitOptions.RemoveEmptyEntries);
+			if (tokens.Length == 0)
+				return last;
+			
+			IPropertyMapping lastEntity = rootPersister;
+			int i = 0;
+			if (entityJoins.TryGetValue(tokens[0], out var entityJoinInfo))
+			{
+				last = entityJoinInfo.Persister;
+				lastEntity = (IPropertyMapping) last;
+				++i;
```

Comment:
`i++;` is more common place, and we do not need pre-increment here.

Reply:
No, of course It's not about optimization :) it's about habit - I just use `++i;` since C++ and it's more common for me.
I have no objections to change it but curious why do you think that `i++` is more common for C#? Is it because VS/Resharper expands `for` with `i++`?

Reply:
P.S. I don't have much time this week but I will try to address all the questions/concerns next week.

Reply:
> Is it because VS/Resharper expands `for` with `i++`?

No. Just because I have far more seen post-increments than pre-increments in C#, and I do not think I am alone. Alexander had made the same demand, switching away from `++i` to `i++` on a previous PR. The opener argued for performance as far a I remember, and it was discarded as not relevant in C#.

Reply:
This was [here](https://github.com/nhibernate/nhibernate-core/pull/367#discussion_r19515494).